packages/types/types/src/record.ts

```diff
@@ -3,45 +3,102 @@
  * Licensed under the MIT License.
  */
 
+/** An observable key/value collection, such as a Record or Map. */
+export interface IShapeProducer<TKey> {
+    /**
+     * Acquire a reader for this producer's values and implicitly subscribe the consumer
+     * to value change notifications.
+     *
+     * @param consumer - The consumer to be notified of value changes.
+     */
+    open(consumer: IShapeConsumer<TKey>): IShapeReader<TKey>;
+
+    /**
+     * Unsubscribe the given 'consumer' from this producer's change notifications.
+     *
+     * @param consumer - The consumer to unregister from the producer.
+     */
+    close(consumer: IShapeConsumer<TKey>): void;
+}
+
+/** A consumer of change notifications for a key/value collection, such as a Record or Map. */
+export interface IShapeConsumer<TKey> {
+    keyChanged(key: TKey, producer: IShapeProducer<TKey>): void;
+}
+
+/** Capability to read values of a key/value collection. */
+export interface IShapeReader<TKey> {
+    keys(): IterableIterator<TKey>;
+
+    has(key: TKey): boolean;
+
+    readonly size: number;
+
+    /**
+     * A reference to the underlying producer that provides the shape for this reader,
+     * or undefined if the producer is immutable.
+     */
+    readonly producer?: IShapeProducer<TKey>;
+}
+
+export interface IShapeWriter<TKey> {
```

Comment:
Do new keys ever get introduced?

Reply:
Strictly, no...  but you can leave the set of keys open using `IWriter<Record<string, any>>`...

It occurs to me now that I should probably stop saying "records are things with an immutable shape" since I use 'Record<TKey, TValue>' to allow the shape to mutate.  Have to come up with another term...